rust/kcl-lib/src/execution/kcl_value.rs

```diff
@@ -594,6 +570,215 @@ impl KclValue {
+                _ => None,
+            },
+            PrimitiveType::Plane => match value {
+                KclValue::Plane { .. } => Some(value.clone()),
+                KclValue::Object { value, meta } => {
+                    let origin = value.get("origin").and_then(Point3d::from_kcl_val)?;
+                    let x_axis = value.get("xAxis").and_then(Point3d::from_kcl_val)?;
+                    let y_axis = value.get("yAxis").and_then(Point3d::from_kcl_val)?;
+                    let z_axis = value.get("zAxis").and_then(Point3d::from_kcl_val)?;
+
+                    let id = exec_state.global.id_generator.next_uuid();
+                    let plane = Plane {
+                        id,
+                        artifact_id: id.into(),
+                        origin,
+                        x_axis,
+                        y_axis,
+                        z_axis,
+                        value: super::PlaneType::Uninit,
+                        // TODO use length unit from origin
+                        units: exec_state.length_unit(),
+                        meta: meta.clone(),
+                    };
+
+                    Some(KclValue::Plane { value: Box::new(plane) })
+                }
+                _ => None,
+            },
+            PrimitiveType::ImportedGeometry => match value {
+                KclValue::ImportedGeometry { .. } => Some(value.clone()),
+                _ => None,
+            },
+        }
+    }
+
+    fn coerce_to_array_type(&self, ty: &PrimitiveType, len: ArrayLen, exec_state: &mut ExecState) -> Option<KclValue> {
+        match self {
+            KclValue::HomArray { value, ty: aty } => {
+                // TODO could check types of values individually
+                if aty != ty {
+                    return None;
+                }
+
+                let value = match len {
+                    ArrayLen::None => value.clone(),
+                    ArrayLen::NonEmpty => {
+                        if value.is_empty() {
+                            return None;
+                        }
+
+                        value.clone()
+                    }
+                    ArrayLen::Known(n) => {
+                        if n > value.len() {
+                            return None;
+                        }
+
+                        value[..n].to_vec()
+                    }
+                };
+
+                Some(KclValue::HomArray { value, ty: ty.clone() })
+            }
+            KclValue::MixedArray { value, .. } => {
+                let value = match len {
+                    ArrayLen::None => value.clone(),
+                    ArrayLen::NonEmpty => {
+                        if value.is_empty() {
+                            return None;
+                        }
+
+                        value.clone()
+                    }
+                    ArrayLen::Known(n) => {
+                        if n > value.len() {
```

Comment:
Should we check `n != value.len()` instead here?

Reply:
Hmm, that's a good question! I was thinking of this in terms of a flexible coercion, so for example if we required an array of 10 elements and got an array of the same thing, but 20, then it's fine. But I guess we don't want to accept a 3d point where the intention is taking a 2d point without a projection, and that is likely to be more common